---
title: "reprogramming_example"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{reprogramming_example}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```


### Load neccecary packages
```{r eval=FALSE}
library(Seurat)
library(ggplot2)
library(slingshot)
library(RColorBrewer)
library(GeneSwitches)
library(GeneSwitchesScorer)
library(SingleCellExperiment)
```


# Data
## Here we use this Reprogramming dataset as the reference 
```{r eval=FALSE}
big_reference.seu <- readRDS("~/R Packages/mini_data/readySeu_hrpi.rds")
```
### Downsample if neccecary
```{r eval=FALSE}
reference.seu <- subset(x = big_reference.seu, downsample = 100)
```

#View the reference
```{r eval=FALSE}
DimPlot(object = reference.seu, reduction = "umap", group.by = "identity" ,label = T) +
  ggtitle("Reference")
```

# Subset the reference to be a single biological trajectory
# in this case fibroblast to primed.
```{r eval=FALSE}
reference.seu <- subset(x = reference.seu, subset = identity %in% grep("^n", reference.seu@meta.data$identity, value = TRUE), invert = TRUE)
```

#View the reduced reference
```{r eval=FALSE}
DimPlot(object = reference.seu, reduction = "umap", group.by = "identity" ,label = T) +
  ggtitle("Reference")
```

## We use subsets of the Reprogramming dataset as queries.
```{r eval=FALSE}
fibroblast.seu <- subset(x = reference.seu, subset = identity %in% "fibroblast")
mixed.seu <- subset(x = reference.seu, subset = identity %in% "mixed")
early_primed.seu <- subset(x = reference.seu, subset = identity %in% "early-primed")
primed.seu <- subset(x = reference.seu, subset = identity %in% "primed")
```


## label transfer
#### **(not neccecary here)**

## re integration
#### **(not neccecary here)**

## Convert objects to single cell experiment objects
```{r eval=FALSE}
reference.sce    <- as.SingleCellExperiment(reference.seu)
fibroblast.sce   <- as.SingleCellExperiment(fibroblast.seu)
mixed.sce        <- as.SingleCellExperiment(mixed.seu)
early_primed.sce <- as.SingleCellExperiment(early_primed.seu)
primed.sce       <- as.SingleCellExperiment(primed.seu)
```


## Run slingshot on the reference data to produce a reprogramming trajectory.
```{r eval=FALSE}
reference.sce  <- slingshot(reference.sce,
                            clusterLabels = "identity",
                            start.clus  = "fibroblast",
                            end.clus = "primed",
                            reducedDim = "UMAP")
```

## Plot the slingshot trajectory.
```{r eval=FALSE}
colors <- colorRampPalette(brewer.pal(11,'Spectral')[-6])(100)
plotcol <- colors[cut(reference.sce$slingPseudotime_1, breaks=100)]
plot(reducedDims(reference.sce)$UMAP, col = plotcol, pch=16, asp = 1)
lines(SlingshotDataSet(reference.sce), lwd=2, col='black')
```

## Convert SCE obj to GS obj
```{r eval=FALSE}
reference.gs    <- SingleCellExperiment(assays = List(expdata = logcounts(reference.sce)))
  colData(reference.gs)$Pseudotime <- reference.sce$slingPseudotime_1
  reducedDims(reference.gs)        <- SimpleList(UMAP = reducedDim(reference.sce, "UMAP", withDimnames=TRUE))

fibroblast.gs   <- SingleCellExperiment(assays = List(expdata = logcounts(fibroblast.sce)))
mixed.gs        <- SingleCellExperiment(assays = List(expdata = logcounts(mixed.sce)))
early_primed.gs <- SingleCellExperiment(assays = List(expdata = logcounts(early_primed.sce)))
primed.gs       <- SingleCellExperiment(assays = List(expdata = logcounts(primed.sce)))
```

##  Choose a Binerization cutoff
##### **this would be good to automate, 
##### or learn how to use the other method of binarizing
```{r eval=FALSE}
 h <- hist(as.matrix(assays(primed.gs)$expdata),
           breaks = 800,
           plot = FALSE)
 plot(h,
      freq = FALSE,
    xlim = c(0.3, 0.9),
    ylim = c(0, 0.5),
      main = "Histogram of gene expression in Reprogramming Study",
      xlab = "Gene expression",
      col = "darkgoldenrod2",
      border = "grey") +
 abline(v = 0.46, col = "blue")
```

## Binarize
```{r eval=FALSE}
reference.gs    <- binarize_exp(reference.gs, fix_cutoff = TRUE, binarize_cutoff = 0.65)
fibroblast.gs   <- binarize_exp(fibroblast.gs, fix_cutoff = TRUE, binarize_cutoff = 0.65)
mixed.gs        <- binarize_exp(mixed.gs, fix_cutoff = TRUE, binarize_cutoff = 0.75)
early_primed.gs <- binarize_exp(early_primed.gs, fix_cutoff = TRUE, binarize_cutoff = 0.49)
primed.gs       <- binarize_exp(primed.gs, fix_cutoff = TRUE, binarize_cutoff = 0.4555)
```

## CHECKPOINT
#### **with how long binerizing takes you may want to choose to save objects for future use.
```{r eval=FALSE}
 saveRDS(reference.gs   , "~/R Packages/mini_data/Binerized/reference.rds")
 saveRDS(fibroblast.gs  , "~/R Packages/mini_data/Binerized/fibroblast.rds")
 saveRDS(mixed.gs       , "~/R Packages/mini_data/Binerized/mixed.rds")
 saveRDS(early_primed.gs, "~/R Packages/mini_data/Binerized/early_primed.rds")
 saveRDS(primed.gs      , "~/R Packages/mini_data/Binerized/primed.rds")

 reference.gs    <- readRDS("~/R Packages/mini_data/Binerized/reference.rds")
 fibroblast.gs   <- readRDS("~/R Packages/mini_data/Binerized/fibroblast.rds")
 mixed.gs        <- readRDS("~/R Packages/mini_data/Binerized/mixed.rds")
 early_primed.gs <- readRDS("~/R Packages/mini_data/Binerized/early_primed.rds")
 primed.gs       <- readRDS("~/R Packages/mini_data/Binerized/primed.rds")
```


### fit logistic regression and find the switching pseudo-time point for each gene
```{r eval=FALSE}
reference_glm.gs <- find_switch_logistic_fastglm(reference.gs, downsample = FALSE, show_warning = FALSE)
```

### **This is another time consuming process, may want to choose to save objects for future use.**
```{r eval=FALSE}
saveRDS(reference_glm.gs   , "~/R Packages/mini_data/Binerized/reference_glm.rds")
reference_glm.gs    <- readRDS("~/R Packages/mini_data/Binerized/reference_glm.rds")
```


## Filter to only include Switching Genes
```{r eval=FALSE}
reference.gs <- filter_switchgenes(reference_glm.gs, allgenes = TRUE,r2cutoff = 0.03)
```

## View all of the switching genes
```{r eval=FALSE}
plot_timeline_ggplot(reference.gs, timedata = colData(reference_glm.gs)$Pseudotime, txtsize = 3)
```

# Filter for an even distribution
```{r eval=FALSE}
gss_genes <- select_evenly_distributed_switching_genes(reference.gs, min_time_spacing = 5)
```

#View the selected switching genes
```{r eval=FALSE}
plot_timeline_ggplot(gss_genes, timedata = colData(reference_glm.gs)$Pseudotime, txtsize = 3)
```

## Reduce the binary counts matricies of the query data to only include the selection of evenly distributed genes from the refernence.
```{r eval=FALSE}
fibroblast_reduced      <- filter_gene_expression_for_switching_genes(fibroblast.gs@assays@data@listData$binary   , gss_genes)
mixed_reduced           <- filter_gene_expression_for_switching_genes(mixed.gs@assays@data@listData$binary        , gss_genes)
early_primed_reduced    <- filter_gene_expression_for_switching_genes(early_primed.gs@assays@data@listData$binary , gss_genes)
primed_reduced          <- filter_gene_expression_for_switching_genes(primed.gs@assays@data@listData$binary       , gss_genes)
```


## Produce an estimate for the position on trajectory of each gene in each cell of a sample.
```{r eval=FALSE}
fibroblast_lines   <- create_racing_lines(fibroblast_reduced  , gss_genes)
mixed_lines        <- create_racing_lines(mixed_reduced       , gss_genes)
early_primed_lines <- create_racing_lines(early_primed_reduced, gss_genes)
primed_lines       <- create_racing_lines(primed_reduced      , gss_genes)
```


## aggregate the above to estimate the position of the query's along the trajectory.
```{r eval=FALSE}
fibroblast_flat   <- flatten_cell_frequencies_moi(fibroblast_lines)
mixed_flat        <- flatten_cell_frequencies_moi(mixed_lines)
early_primed_flat <- flatten_cell_frequencies_moi(early_primed_lines)
primed_flat       <- flatten_cell_frequencies_moi(primed_lines)

plot(colSums(fibroblast_flat))
plot(colSums(mixed_flat))
plot(colSums(early_primed_flat))
plot(colSums(primed_flat))
```


##plot something better.
```{r eval=FALSE}
plot(x = 1:ncol(fibroblast_flat), y = fibroblast_flat[1,], type = "l", xlab = "Pseudotime Index", ylab = "Cell Position Likelyhood", main = "Trajectory Progress of \"fibroblast\"")
plot(x = 1:ncol(mixed_flat), y = mixed_flat[1,], type = "l", xlab = "Pseudotime Index", ylab = "Cell Position Likelyhood", main = "Trajectory Progress of mixed")
plot(x = 1:ncol(early_primed_flat), y = early_primed_flat[1,], type = "l", xlab = "Pseudotime Index", ylab = "Cell Position Likelyhood", main = "Trajectory Progress of early_primed")
plot(x = 1:ncol(primed_flat), y = primed_flat[1,], type = "l", xlab = "Pseudotime Index", ylab = "Cell Position Likelyhood", main = "Trajectory Progress of \"primed\"")
```


## is owens method any different?
```{r eval=FALSE}
fibroblast_flat   <- flatten_cell_frequencies_owen(fibroblast_lines)
mixed_flat        <- flatten_cell_frequencies_owen(mixed_lines)
early_primed_flat <- flatten_cell_frequencies_owen(early_primed_lines)
primed_flat       <- flatten_cell_frequencies_owen(primed_lines)
```
